test/TEST-38-FREEZER/testsuite.sh

```diff
@@ -0,0 +1,308 @@
+#!/usr/bin/env bash
+
+set -e
+set -o pipefail
+
+systemd-analyze log-level debug
+systemd-analyze log-target console
+
+unit=foo.service
+unit_file=/etc/systemd/system/${unit}
+
+setup_test_service() {
+    cat > ${unit_file} <<EOF
+[Service]
+ExecStart=/bin/sleep 3600
+EOF
+
+    systemctl daemon-reload
+    systemctl start "${unit}"
+}
+
+dbus_freeze() {
+    local suffix=
+    suffix="${1##*.}"
+
+    local name="${1%.$suffix}"
+    local object_path="/org/freedesktop/systemd1/unit/${name}_2e${suffix}"
+
+    busctl call \
+           org.freedesktop.systemd1 \
+           "${object_path}" \
+           org.freedesktop.systemd1.Unit \
+           Freeze
+}
+
+dbus_thaw() {
+    local suffix=
+    suffix="${1##*.}"
+
+    local name="${1%.$suffix}"
+    local object_path="/org/freedesktop/systemd1/unit/${name}_2e${suffix}"
+
+    busctl call \
+           org.freedesktop.systemd1 \
+           "${object_path}" \
+           org.freedesktop.systemd1.Unit \
+           Thaw
+}
+
+dbus_freeze_unit() {
+    busctl call \
+           org.freedesktop.systemd1 \
+           /org/freedesktop/systemd1 \
+           org.freedesktop.systemd1.Manager \
+           FreezeUnit \
+           s \
+           "$1"
+}
+
+dbus_thaw_unit() {
+    busctl call \
+           org.freedesktop.systemd1 \
+           /org/freedesktop/systemd1 \
+           org.freedesktop.systemd1.Manager \
+           ThawUnit \
+           s \
+           "$1"
+}
+
+dbus_can_freeze() {
+    local suffix=
+    suffix="${1##*.}"
+
+    local name="${1%.*}"
+    local object_path="/org/freedesktop/systemd1/unit/${name}_2e${suffix}"
+
+    busctl get-property \
+           org.freedesktop.systemd1 \
+           "${object_path}" \
+           org.freedesktop.systemd1.Unit \
+           CanFreeze
+}
+
+check_freezer_state() {
+    local suffix=
+    suffix=$(get_suffix "$1")
+
+    local name="${1%.$suffix}"
+    local object_path="/org/freedesktop/systemd1/unit/${name}_2e${suffix}"
+
+    state=$(busctl get-property \
+                   org.freedesktop.systemd1 \
+                   "${object_path}" \
+                   org.freedesktop.systemd1.Unit \
+                   FreezerState | cut -d " " -f2 | tr -d '"')
+
+    [ "$state" = "$2" ] || {
+        echo "error: unexpected freezer state, expected: $2, actual: $state" >&2
+        exit 1
+    }
+}
+
+check_cgroup_state() {
+    grep -q "frozen $2" /sys/fs/cgroup/system.slice/"$1"/cgroup.events
+}
+
+cleanup() {
+    set +e
+    systemctl stop "${unit}" >/dev/null 2>&1
+    rm -f "${unit_file}"
+    systemctl daemon-reload
+}
```

Comment:
This cleanup is not needed. TEST-* doesn't have to leave the machine in pristine state (as long as everything goes away after after a reboot).